daml-assistant/daml-helper/src/DamlHelper.hs

```diff
@@ -346,45 +324,127 @@ runInit targetFolderM = do
+                , ""
+                , "    daml new myproject " <> projectName
+                , ""
+                ]
+            exitFailure
+
+    -- Ensure we are not creating a project inside another project.
+    targetFolderAbs <- makeAbsolute targetFolder
+
+    damlRootM <- findDamlProjectRoot targetFolderAbs
+    whenJust damlRootM $ \damlRoot -> do
+        hPutStr stderr $ unlines
+            [ "Target directory is inside existing DAML project " <> show damlRoot
+            , "Please specify a new directory outside an existing project."
+            ]
+        exitFailure
+
+    daRootM <- findDaProjectRoot targetFolderAbs
+    whenJust daRootM $ \daRoot -> do
+        hPutStr stderr $ unlines
+            [ "Target directory is inside existing DA project " <> show daRoot
+            , "Please convert DA project to DAML using 'daml init':"
+            , ""
+            , "    daml init " <> escapePath daRoot
+            , ""
+            , "Or specify a new directory outside an existing project."
+            ]
+        exitFailure
+
+    -- Copy the template over.
     copyDirectory templateFolder targetFolder
     files <- listFilesRecursive targetFolder
     mapM_ setWritable files
 
-    -- update daml.yaml
+    -- Update daml.yaml
     let configPath = targetFolder </> projectConfigName
         configTemplatePath = configPath <.> "template"
 
     whenM (doesFileExist configTemplatePath) $ do
         configTemplate <- readFileUTF8 configTemplatePath
         sdkVersion <- getSdkVersion
-        let projectName = takeFileName (dropTrailingPathSeparator targetFolder)
-            config = replace "__VERSION__"  sdkVersion
+        let config = replace "__VERSION__"  sdkVersion
                    . replace "__PROJECT_NAME__" projectName
                    $ configTemplate
         writeFileUTF8 configPath config
         removeFile configTemplatePath
 
+    -- Done.
     putStrLn $
         "Created a new project in \"" <> targetFolder <>
         "\" based on the template \"" <> templateName <> "\"."
 
+defaultProjectTemplate :: String
+defaultProjectTemplate = "skeleton"
+
+legacyConfigName :: FilePath
+legacyConfigName = "da.yaml"
+
+findDamlProjectRoot :: FilePath -> IO (Maybe FilePath)
+findDamlProjectRoot = findAscendantWithFile projectConfigName
+
+findDaProjectRoot :: FilePath -> IO (Maybe FilePath)
+findDaProjectRoot = findAscendantWithFile legacyConfigName
+
+findAscendantWithFile :: FilePath -> FilePath -> IO (Maybe FilePath)
+findAscendantWithFile filename path =
+    findM (\p -> doesFileExist (p </> filename)) (ascendants path)
+
+-- | Escape special characters in a filepath so they can be used as a shell
+-- argument when displaying a suggested command to user. Do not use this to
+-- invoke shell commands directly (there are libraries designed for that).
+escapePath :: FilePath -> FilePath
+escapePath p | isWindows = concat ["\"", p, "\""] -- Windows is a mess
+escapePath p = p >>= \c ->
```

Comment:
I would probably just enclose it in `'` and ignore the case where `'` is in the project name but your choice :smile:

Reply:
I find it annoying when tools add quotation marks around paths when they aren't strictly necessary ... so I'll stick with this for now.